File: tests/arch/fixtures/real_world/php/src/Result/Failure.php
<?php

declare(strict_types=1);

namespace Example\Result;

use Example\Result\IResult;

/**
 * Represents a failed result carrying an error message.
 */
final class Failure implements IResult
{
    public function __construct(private readonly string $error) {}

    public function isSuccess(): bool
    {
        return false;
    }

    /** @return null */
    public function getValue(): mixed
    {
        return null;
    }

    public function getError(): ?string
    {
        return $this->error;
    }

    public function __toString(): string
    {
        return sprintf('Failure(%s)', $this->error);
    }
}

File: tests/arch/fixtures/real_world/csharp/src/Result/Failure.cs
using Example.Result.IResult;

namespace Example.Result;

/// <summary>Represents a failed result carrying an error message.</summary>
public sealed class Failure<T> : IResult<T>
{
    public Failure(string error) => Error = error;

    public bool IsSuccess => false;
    public T? Value => default;
    public string Error { get; }

    public override string ToString() => $"Failure({Error})";
}
